crates/circuit/src/packed_instruction.rs

```diff
@@ -0,0 +1,497 @@
+    const DISCRIMINANT_MASK: usize = 0b11;
+    /// The number of bits used to store the discriminant metadata.
+    const DISCRIMINANT_BITS: u32 = Self::DISCRIMINANT_MASK.count_ones();
+    /// A bitmask that masks out only the standard gate information.  This should always have the
+    /// same effect as `POINTER_MASK` because the high bits should be 0 for a `StandardGate`, but
+    /// this is defensive against us adding further metadata on `StandardGate` later.  After
+    /// masking, the resulting integer still needs shifting downwards to retrieve the standard gate.
+    const STANDARD_GATE_MASK: usize = (u8::MAX as usize) << Self::DISCRIMINANT_BITS;
+    /// A bitmask that retrieves the stored pointer directly.  The discriminant is stored in the
+    /// low pointer bits that are guaranteed to be 0 by alignment, so no shifting is required.
+    const POINTER_MASK: usize = usize::MAX ^ Self::DISCRIMINANT_MASK;
+
+    /// Extract the discriminant of the operation.
+    #[inline]
+    fn discriminant(&self) -> PackedOperationType {
+        ::bytemuck::checked::cast((self.0 & Self::DISCRIMINANT_MASK) as u8)
+    }
+
+    /// Get the contained pointer to the `PyGate`/`PyInstruction`/`PyOperation` that this object
+    /// contains.
+    ///
+    /// **Panics** if the object represents a standard gate; see `try_pointer`.
+    #[inline]
+    fn pointer(&self) -> NonNull<()> {
+        self.try_pointer().unwrap()
+    }
+
+    /// Get the contained pointer to the `PyGate`/`PyInstruction`/`PyOperation` that this object
+    /// contains.
+    ///
+    /// Returns `None` if the object represents a standard gate.
+    #[inline]
+    pub fn try_pointer(&self) -> Option<NonNull<()>> {
+        match self.discriminant() {
+            PackedOperationType::StandardGate => None,
+            PackedOperationType::Gate
+            | PackedOperationType::Instruction
+            | PackedOperationType::Operation => {
+                let ptr = (self.0 & Self::POINTER_MASK) as *mut ();
+                // SAFETY: `PackedOperation` can only be constructed from a pointer via `Box`, which
+                // is always non-null (except in the case that we're partway through a `Drop`).
+                Some(unsafe { NonNull::new_unchecked(ptr) })
+            }
+        }
+    }
+
+    /// Get the contained `StandardGate`.
+    ///
+    /// **Panics** if this `PackedOperation` doesn't contain a `StandardGate`; see
+    /// `try_standard_gate`.
+    #[inline]
+    pub fn standard_gate(&self) -> StandardGate {
+        self.try_standard_gate().unwrap()
+    }
+
+    /// Get the contained `StandardGate`, if any.
+    #[inline]
+    pub fn try_standard_gate(&self) -> Option<StandardGate> {
+        match self.discriminant() {
+            PackedOperationType::StandardGate => ::bytemuck::checked::try_cast(
+                ((self.0 & Self::STANDARD_GATE_MASK) >> Self::DISCRIMINANT_BITS) as u8,
+            )
+            .ok(),
```

Comment:
Following the same ideas as above, you could write
```suggestion
            PackedOperationType::StandardGate => Some(
                ::bytemuck::checked::try_cast(
                    ((self.0 & Self::STANDARD_GATE_MASK) >> Self::DISCRIMINANT_BITS) as u8,
                )
                .expect("Qiskit bug: Discriminant encodes no StandardGate"),
```
Again, an argument is that this helps the reader of the code even if the bug is never triggered.

Reply:
imo putting these `expect`s with messages that just say the same thing as the code is making the call sites noisier without helping debugging.  The standard panic includes the file and line number, so we can immediately track down the failing line, and this line of code clearly describes the same as the error (actually, even more than the message).

imo a good error message in one of these would say _why_ the assumptions are the way they are, not re-describe _what_ failed.  Here, the assumption is the basically the entire data representation, so I don't think it's especially helpful to add the noise.

Reply:
I disagree. If it were up to me I'd make these changes or something else along these lines.  But I've already stated my arguments.

> imo a good error message in one of these would say why the assumptions 

I don't really follow, but if you think there's a better error message, that would be good. The messages I wrote are not the best or only choices.

I assumed that you wrote the code the way you wanted and wouldn't want to change it. But I thought it's important to make a record of my preference (it's not an uncommon preference). I don't judge it important enough for us to spend the effort debating it at present. Better to leave all of the code I commented on as is and move on.

Reply:
Just to clarify the "why" thing: consider a simpler example of our OQ3 importer hypothetically wanting to read and cache our vendored `stdgates.inc` file when the module initialises. We vendor it, so the file absolutely should exist. In that case, I'd write `File::open(...).expect("'stdgates.inc' should be vendored with Qiskit")` or similar. The panic still shows the line and the `Err` variant data (which is the _what_), but the message now explains the programmer's intent as to _why_ they expected it to work.

Reply:
So this is an example of when you *would* want to use expect.

Reply:
Actually, looking back - this particular one isn't even an `unwrap`, it's an `ok`.  That's not a panicking method - it just turns a `Result<T, _>` into an `Option<T>` by erasure of any error variant, which is what this method is doing.

And yeah, I'd `expect` if a file failed to open that we knew should exist.